examples/tls_rustls_dynamic_certs.rs

```diff
@@ -0,0 +1,218 @@
+//! ```
+//!
+//! Second example certificate:
+//! ```sh
+//! curl -vik --resolve second.example:64802:127.0.0.1 https://second.example:64802
+//! ```
+//! Output
+//! ```
+//! * Server certificate:
+//! *  subject: CN=second.example.com
+//! *  start date: Dec  9 20:08:11 2024 GMT
+//! *  expire date: Dec  7 20:08:11 2034 GMT
+//! *  issuer: CN=second.example.com
+//! *  SSL certificate verify result: self signed certificate (18), continuing anyway.
+//! ```
+//!
+//! Fallback to to default (example certificate) if no matches are found:
+//! ```sh
+//! curl -vik https://127.0.0.1:64802
+//! ```
+//! Output
+//! ```
+//! * Server certificate:
+//! *  subject: CN=example.com
+//! *  start date: Dec  9 20:05:17 2024 GMT
+//! *  expire date: Dec  7 20:05:17 2034 GMT
+//! *  issuer: CN=example.com
+//! *  SSL certificate verify result: self signed certificate (18), continuing anyway.
+//! ```
+
+// rama provides everything out of the box to build a TLS termination proxy
+use rama::{
+    Context, Layer,
+    error::{ErrorContext, OpaqueError},
+    graceful::Shutdown,
+    http::{IntoResponse, Request, Response, server::HttpServer},
+    layer::ConsumeErrLayer,
+    net::{address::Domain, tls::ApplicationProtocol, tls::client::ClientHello},
+    rt::Executor,
+    service::service_fn,
+    tcp::server::TcpListener,
+    tls::rustls::{
+        dep::{
+            pemfile,
+            rustls::{
+                ALL_VERSIONS, ServerConfig, crypto::aws_lc_rs, server::ResolvesServerCert,
+                sign::CertifiedKey,
+            },
+        },
+        server::{TlsAcceptorDataBuilder, TlsAcceptorLayer},
+    },
+};
+
+// everything else is provided by the standard library, community crates or tokio
+use std::{convert::Infallible, io::BufReader, sync::Arc, time::Duration};
+use tracing::metadata::LevelFilter;
+use tracing_subscriber::{EnvFilter, fmt, prelude::*};
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::DEBUG.into())
+                .from_env_lossy(),
+        )
+        .init();
+
+    let shutdown = Shutdown::default();
+
+    // Dynamic certs using [`ResolvesServerCert`] in rustls are loaded synchronously, if you need to
+    // load them in an async way, see example `tls_rustls_dynamic_config` for how to provide a different
+    // [`rustls::ServerConfig`] depending on received client_hello in an async context (TODO will be added this week)
+
+    let dynamic_issuer = Arc::new(DynamicIssuer::new());
+    let config = ServerConfig::builder_with_protocol_versions(ALL_VERSIONS)
+        .with_no_client_auth()
+        .with_cert_resolver(dynamic_issuer);
+
+    let acceptor_data = TlsAcceptorDataBuilder::from(config)
+        .with_alpn_protocols(&[ApplicationProtocol::HTTP_11, ApplicationProtocol::HTTP_2])
```

Comment:
you might btw want to add convenience methods for the common case. E.g. `with_alpn_protocols_http_auto`, as that is probably the most common